runtime/common/src/xcm.rs

```diff
@@ -158,62 +154,76 @@ where
 /// `MultiLocation`.
 pub struct CurrencyIdConvert<T>(PhantomData<T>);
 
+impl<T> MaybeEquivalence<MultiLocation, CurrencyId> for CurrencyIdConvert<T>
+where
+	T: orml_asset_registry::Config<AssetId = CurrencyId, CustomMetadata = CustomMetadata>
+		+ parachain_info::Config,
+{
+	fn convert(location: &MultiLocation) -> Option<CurrencyId> {
+		let para_id = parachain_info::Pallet::<T>::parachain_id();
+		let unanchored_location = match location {
+			MultiLocation {
+				parents: 0,
+				interior,
+			} => MultiLocation {
+				parents: 1,
+				interior: interior
+					.pushed_front_with(Parachain(u32::from(para_id)))
+					.ok()?,
+			},
+			x => *x,
+		};
+
+		orml_asset_registry::Pallet::<T>::location_to_asset_id(unanchored_location)
+	}
+
+	fn convert_back(id: &CurrencyId) -> Option<MultiLocation> {
+		orml_asset_registry::Pallet::<T>::metadata(id)
+			.filter(|m| m.additional.transferability.includes_xcm())
+			.and_then(|m| m.location)
+			.and_then(|l| l.try_into().ok())
+	}
+}
+
 /// Convert our `CurrencyId` type into its `MultiLocation` representation.
 /// We use the `AssetRegistry` to lookup the associated `MultiLocation` for
 /// any given `CurrencyId`, while blocking tokens that are not Xcm-transferable.
 impl<T> Convert<CurrencyId, Option<MultiLocation>> for CurrencyIdConvert<T>
 where
-	T: orml_asset_registry::Config<AssetId = CurrencyId, CustomMetadata = CustomMetadata>,
+	T: orml_asset_registry::Config<AssetId = CurrencyId, CustomMetadata = CustomMetadata>
+		+ parachain_info::Config,
 {
 	fn convert(id: CurrencyId) -> Option<MultiLocation> {
-		orml_asset_registry::Pallet::<T>::metadata(&id)
-			.filter(|m| m.additional.transferability.includes_xcm())
-			.and_then(|m| m.location)
-			.and_then(|l| l.try_into().ok())
+		<Self as MaybeEquivalence<_, _>>::convert_back(&id)
 	}
 }
 
 /// Convert an incoming `MultiLocation` into a `CurrencyId` through a
 /// reverse-lookup using the AssetRegistry. In the registry, we register CFG
-/// using its absolute, non-anchored MultliLocation so we need to unanchor the
+/// using its absolute, non-anchored MultiLocation so we need to unanchor the
 /// input location for Centrifuge-native assets for that to work.
-impl<T> xcm_executor::traits::Convert<MultiLocation, CurrencyId> for CurrencyIdConvert<T>
+impl<T> Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert<T>
```

Comment:
Is the converter forced to only optionally convert from location to currency?

Reply:
Yes, it's forced by the users of this type though this trait implementations.

Also now `xcm_executor::traits::Convert` no longer exists (note that `Convert` is from `sp_runtime` which is a different type).